plugins/catalog-backend/src/ingestion/processors/CodeOwnersProcessor.ts

```diff
@@ -0,0 +1,172 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity, LocationSpec } from '@backstage/catalog-model';
+import {
+  LocationProcessor,
+  LocationProcessorEmit,
+  LocationProcessorRead,
+} from './types';
+import * as codeowners from 'codeowners-utils';
+import { CodeOwnersEntry } from 'codeowners-utils';
+import parseGitUri from 'git-url-parse';
+import { filter, head, get, pipe, reverse } from 'lodash/fp';
+
+// NOTE: This can be removed when ES2021 is implemented
+import 'core-js/features/promise';
+
+const ALLOWED_LOCATION_TYPES = [
+  'azure/api',
+  'bitbucket/api',
+  'github',
+  'github/api',
+  'gitlab',
+  'gitlab/api',
+];
+
+export class CodeOwnersProcessor implements LocationProcessor {
+  async processEntity(
+    entity: Entity,
+    location: LocationSpec,
+    _emit: LocationProcessorEmit,
+    read: LocationProcessorRead,
+  ): Promise<Entity> {
+    // Only continue if the owner is not set
+    if (
+      !entity ||
+      !['Component', 'API'].includes(entity.kind) ||
+      !ALLOWED_LOCATION_TYPES.includes(location.type) ||
+      (entity.spec && entity.spec.owner)
+    ) {
+      return entity;
+    }
+
+    const owner = await resolveCodeOwner(location, read);
+
+    return {
+      ...entity,
+      spec: { ...entity.spec, owner },
+    };
+  }
+}
+
+export async function resolveCodeOwner(
+  location: LocationSpec,
+  read: LocationProcessorRead,
+): Promise<string | undefined> {
+  const ownersText = await findRawCodeOwners(location, read);
+
+  if (!ownersText) {
+    throw Error(`Unable to find codeowners file for: ${location.target}`);
+  }
+
+  const owners = parseCodeOwners(ownersText);
+
+  return findPrimaryCodeOwner(owners);
+}
+
+export async function findRawCodeOwners(
+  location: LocationSpec,
+  read: LocationProcessorRead,
+): Promise<string | undefined> {
+  const readOwnerLocation = async (basePath: string): Promise<string> => {
+    const ownerLocation = buildCodeOwnerLocation(
+      location,
+      `${basePath}/CODEOWNERS`,
+    );
+
+    return readLocation(ownerLocation, read);
+  };
+
+  const gitProvider = location.type.split('/')[0];
+
+  return Promise.any([
+    readOwnerLocation(`/.${gitProvider}`),
+    readOwnerLocation(''),
+    readOwnerLocation('/docs'),
+  ]);
+}
+
+export function parseCodeOwners(ownersText: string) {
+  return codeowners.parse(ownersText);
+}
+
+export function findPrimaryCodeOwner(
+  owners: CodeOwnersEntry[],
+  pattern = '*',
+): string | undefined {
+  return pipe(
+    filter((e: CodeOwnersEntry) => e.pattern === pattern),
```

Comment:
Shouldn't this always be `e.pattern === '*' || urlOfTheYamlFile.startsWith(e.pattern)`, and give urlOfTheYamlFile as a parameter to the function?

Reply:
I was thinking ahead to support a nested component entity in a monorepo setup. It's not implemented here as it's not completely trivial to match the locationspec subpath to a gitignore style codeowners entry. So therefore, I left it out. In the meantime, I can hard code it to `*` though.

Reply:
But the rules of the codeowners file is strictly that the last matching entry by prefix wins, monorepo or not, therefore the order of the lines is significant. So that detail I am sure we cannot miss out on. In either case, the `===` check I think isn't correct, it has to be a startsWith essentially.

Reply:
In the sake of getting a first cut out faster, I've just removed the handling for sub paths for now. Figured this could be added in another pass.